compiler/rustc_data_structures/src/owning_ref/mod.rs

```diff
@@ -1117,41 +1111,41 @@ impl<T: 'static> ToHandleMut for RefCell<T> {
 // about which handle creation to use (i.e., read() vs try_read()) as well as
 // what to do with error results.
 
-/// Typedef of an owning reference that uses a `Box` as the owner.
+/// Type alias of an owning reference that uses a [`Box`] as the owner.
```

Comment:
While we're at it, could you delete this impl for Box? It's unsound. If it's used we need to find a replacement. We can probably just use Arc without regressing perf.

Reply:
What impl? This is a type alias...

Reply:
Maybe the impl lives in the stable_deref_trait crate. But at least the type alias should be deleted, we should absolutely not be using an owning_ref with a Box, that's unsound because of boxes aliasing guarantees.

Reply:
Would it be okay if we wrap the owner in `MaybeUninit` to hide the aliasing? 🤔 

At least it convinces the miri I think: 
- [Current version](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=50ba5fb85fbc7f7033665247f8b44087) (miri detects ub ❌)
- [With `MaybeUninit`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2e2cea217d942b48646857574c6e21e8) (miri doesn't see wrong ✅)

Although I'm not sure if this is just a bug in MIRI...

Also/1: why do we have a copy of the `owning_ref` crate? 🤔 
Also/2: wait `owning_ref` crate (same as with this module) is straight up unsound? 🤔

Reply:
MaybeUninit would work but that'd lose the niche and it's also kinda hacky. I'd prefer just deleting the Box and using an Arc, it's the easiest solution. If we really want a Box, then we can roll or own or use the one from ouroboros. Or we can make T-opsem give up on giving Box these bad and confusing rules :D

I think owning_ref had some unrelated unsoundnesses and is unmaintained which is why the fixed version is copied here.

Reply:
I've searched for uses of `(Owning|Box)Ref` and we appear to only have 3 uses:
- With `Mmap` (which is itself unsound 💀 ) in `rustc_codegen_ssa::back::metadata::DefaultMetadataLoader` (uses box to erase type)
- With `Mmap` ( 💀 ) and `Vec<u8>` in `rustc_metadata::locator` (uses box to erase type)
- With `MetadataBlob` in `rustc_metadata::rmeta::def_path_hash_map`

Given that all of these use `T = [u8]` and most erase `U`... maybe we can "just" yeet `owning_ref` with all it's unsoundness altogether and instead have something like [this](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=09fc3062485c8607948841b26457243c)?

Reply:
I tried an overcomplicated version of it in #97770. Your example sounds great, you should try it :).